DEVELOPING_OPAL/demos/src/main/scala/org/opalj/tac/fpcf/analyses/taint/ForwardClassForNameTaintAnalysisScheduler.scala

```diff
@@ -0,0 +1,149 @@
+/* BSD 2-Clause License - see OPAL/LICENSE for details. */
+package org.opalj.tac.fpcf.analyses.taint
+
+import org.opalj.br.analyses.{DeclaredMethodsKey, ProjectInformationKeys, SomeProject}
+import org.opalj.br.fpcf.PropertyStoreKey
+import org.opalj.br.{DeclaredMethod, Method, ObjectType}
+import org.opalj.fpcf.{FinalEP, PropertyBounds, PropertyStore}
+import org.opalj.ifds.{IFDSAnalysis, IFDSAnalysisScheduler, IFDSProperty, IFDSPropertyMetaInformation}
+import org.opalj.tac.cg.{RTACallGraphKey, TypeIteratorKey}
+import org.opalj.tac.fpcf.analyses.ifds.taint.{FlowFact, ForwardTaintProblem, TaintFact, TaintProblem, Variable}
+import org.opalj.tac.fpcf.analyses.ifds._
+import org.opalj.tac.fpcf.properties.{TACAI, Taint}
+import org.opalj.tac.fpcf.properties.cg.Callers
+
+import java.io.File
+
+/**
+ * A forward IFDS taint analysis, which tracks the String parameters of all methods of the rt.jar,
+ * which are callable from outside the library, to calls of Class.forName.
+ *
+ * @author Dominik Helm
+ * @author Mario Trageser
+ * @author Michael Eichberg
+ */
+class ForwardClassForNameTaintAnalysis(project: SomeProject)
+    extends IFDSAnalysis()(project, new ForwardClassForNameTaintProblem(project), Taint)
+
+class ForwardClassForNameTaintProblem(project: SomeProject)
+    extends ForwardTaintProblem(project) with TaintProblem[Method, JavaStatement, TaintFact] {
+    private val propertyStore = project.get(PropertyStoreKey)
+    /**
+     * Returns all methods, that can be called from outside the library.
+     * The call graph must be computed, before this method may be invoked.
+     *
+     * @return All methods, that can be called from outside the library.
+     */
+    protected def methodsCallableFromOutside: Set[DeclaredMethod] = {
+        declaredMethods.declaredMethods.filter(canBeCalledFromOutside).toSet
```

Comment:
For the JDK this doesn't seem to change much performance wise. It seems to be slightly slower ~86 ms vs ~92 ms and arguably uglier. Maybe this can be improved, no Option Allocation?
`            propertyStore(declaredMethods.declaredMethods.iterator.to(Iterable), Callers.key).flatMap {
                case FinalEP(method, callers) if callers.hasCallersWithUnknownContext => Some(method)
                case _ => None
            }.toSet
`

Reply:
.collect instead of .flatmap has roughly the same performance

Reply:
collect instead of flatMap should avoid the Option allocation. Not sure if it's better, I agree this is pretty ugly. I thought it would be possible to just pass declaredMethods.declaredMethods to the propertystore, but obviously that only accepts Iterables, so probably not much to gain here.
Why is the `.iterator` needed, this already is an iterator, isn't it?

Reply:
> .collect instead of .flatmap has roughly the same performance

Beat me to it

Reply:
removed the .iterator already.